docs/plotting.md

```diff
@@ -0,0 +1,97 @@
+# Recommended interface and behavior of a `plot` method
+
+Here we recommend the input arguments, return value, and behavior of the
+`plot` method of a class.
+
+## Requirements
+
+1. **Convenience to interactive users.** This is the highest priority.
+    Compared to being called in a batch script as a library (for composing
+    more complicated plots or other purposes), the `plot` method is mainly
+    used in interactive sessions like ipython, jupyther, colab, PyCharm,
+    and python interpreter.
+1. **No unnecessary messages in interactive sessions.** It should not produce
+    any warning/error messages in normal usages in an interactive sessions.
+    See [#1890](https://github.com/quantumlib/Cirq/issues/1890#issue-473510953)
+    for an example of such message.
+1. **No popups during tests.** It should not produce any pop-up windows during
+    tests.
+
+## Recommendation
+
+The `plot` method must produce a plot when there is no arguments in an
+interactive session. The recommended way to achieve that is illustrated in the
+example below.
+
+```python
+class Foo:
+    ...
+    def plot(self, ax: plt.Axes=None, **plot_kwargs: Any) -> plt.Axes:
+        show_plot = not ax
+        if show_plot:
+            fig, ax = plt.subplots(1, 1)  # or your favorite figure setup
+        # Call methods of the ax instance like ax.plot to plot on it.
+        ...
+        if show_plot:
+            fig.show()
+        return ax
+```
+
+This `plot` method works in 2 modes: *memory mode* and *interactive mode*,
+signalled by the presence of the `ax` argument. When present, the method is
+instructed to plot on the provided `ax` instance in memory. No plot is shown
+on the screen. When absent, the code is in *interactive* mode, and it creates
+a figure and shows it.
+
+The returned `ax` instance can be used to further customize the plot if the
+user wants to. Note that if we were to call `plt.show` instead of `fig.show`,
```

Comment:
What happens if we return `None` when `ax` is not set? Pycharm seem to require `plt.show()` instead of `fig.show()`.

Reply:
Once `plt.show()` is called, the figure and ax can no longer be customized (e.g., add grid, change an axis label, etc).

Per offline conversation, the problem is that it doesn't work in the PyCharm environment. The non-blocking `fig.show` causes the pop-up to not be shown. Is it actually shown but disappears immediately? Is there a setting that can change PyCharm's behavior?

Customizability of plots is not in the current requirements, but I think it is very important. I'd like to add it to the list of requirements.

Reply:
It's fine if we don't support pycharm and instead nag their devs to make it work.